File: ruby/isbn-verifier/isbn_verifier.rb
# https://exercism.org/tracks/ruby/exercises/isbn-verifier
class IsbnVerifier
  def self.valid?(isbn)
    digits = isbn.scan(/[\dX]/)
    return false if digits.size != 10

    digits[-1] = 10 if digits[-1] == 'X'

    checksum = 0
    digits.map!(&:to_i).each_index do |i|
      checksum += digits[i] * (10 - i)
    end

    (checksum % 11).zero?
  end
end

File: go/isbn-verifier/isbn_verifier.go
package isbnverifier

func IsValidISBN(isbn string) bool {
	var n, sum, counter int

	for i := 0; i < len(isbn); i++ {
		if counter > 10 {
			return false
		}
		ch := isbn[i]

		switch {
		case ch == '-':
			continue

		case ch == 'X':
			if counter != 9 {
				return false
			}
			n = 10

		case ch >= '0' && ch <= '9':
			n = int(ch - '0')

		default:
			return false
		}

		sum += n * (10 - counter)
		counter++
	}

	return counter == 10 && sum%11 == 0
}
